azext_iot/iothub/providers/device_identity.py

```diff
@@ -0,0 +1,658 @@
+            )
+
+            # delete non-tarred folder
+            rmtree(device_cert_output_directory)
+
+        # Give device registry a chance to catch up
+        sleep(1)
+
+        # Get all device ids and scopes (inconsistent timing, hence sleep)
+        query_args = ["SELECT deviceId, deviceScope FROM devices"]
+        query_method = self.service_sdk.query.get_twins
+        all_hub_devices = _execute_query(query_args, query_method)
+
+        # Throw an error if we don't get the same number of desired devices back
+        if len(all_hub_devices) < len(config.devices):
+            raise AzureResponseError(
+                "An error occurred - Failed to fetch device scopes for all devices"
+            )
+
+        # set all device scopes
+        scope_dict: Dict[str, str] = {}
+        for device in all_hub_devices:
+            id = device["deviceId"]
+            if device_config_dict.get(id, None):
+                scope_dict[id] = device["deviceScope"]
+
+        # Set parent / child relationships
+        device_to_parent_iterator = (
+            tqdm(device_to_parent_dict, desc="Setting device parents")
+            if visualize
+            else device_to_parent_dict
+        )
+        for device_id in device_to_parent_iterator:
+            # get device properties
+            device = self.service_sdk.devices.get_identity(id=device_id)
+            parent_id = device_to_parent_dict[device_id]
+            parent_scope = scope_dict[parent_id]
+            # set new parent scope
+            device.parent_scopes = [parent_scope]
+            # update device
+            self.service_sdk.devices.create_or_update_identity(
+                id=device_id, device=device, if_match="*"
+            )
+
+        # update edge config / set-modules
+        devices_config_iterator = (
+            tqdm(config.devices, desc="Setting edge module content")
+            if visualize
+            else config.devices
+        )
+        for device_config in devices_config_iterator:
+            device_id = device_config.device_id
+            deployment_content = device_config.deployment
+            if deployment_content:
+                self.service_sdk.configuration.apply_on_edge_device(
+                    id=device_id, content=deployment_content
+                )
+
+    def _process_nested_edge_config_file_content(
+        self, content: dict
+    ) -> NestedEdgeConfig:
+        """
+        Process edge config file schema dictionary
+        """
+        # TODO - version / schema validation
+        version = content.get("config_version", None)
+        hub_config = content.get("iothub", None)
+        devices_config = content.get("edgedevices", [])
+        for check, err in [
+            (version, "No schema version specified in configuration file"),
+            (hub_config, "No `iothub` properties specified in configuration file")
+            # (len(devices_config), "No devices specified in configuration file")
```

Comment:
We should probably throw an exception if a config file is used without any devices - going to uncomment this and update tests.